src/circom/r1cs_reader.rs

```diff
@@ -54,21 +57,39 @@ impl<E: PairingEngine> R1CSFile<E> {
             return Err(Error::new(ErrorKind::InvalidData, "Unsupported version"));
         }
 
-        let _num_sections = reader.read_u32::<LittleEndian>()?;
+        let num_sections = reader.read_u32::<LittleEndian>()?;
 
-        // todo: rewrite this to support different section order and unknown sections
         // todo: handle sec_size correctly
-        let _sec_type = reader.read_u32::<LittleEndian>()?;
-        let sec_size = reader.read_u64::<LittleEndian>()?;
+        // section type -> file offset
+        let mut sec_offsets = HashMap::<u32, u64>::new();
+        let mut sec_sizes = HashMap::<u32, u64>::new();
+
+        // get file offset of each section
+        for _ in 0..num_sections {
+            let sec_type = reader.read_u32::<LittleEndian>()?;
+            let sec_size = reader.read_u64::<LittleEndian>()?;
+            let offset = reader.seek(SeekFrom::Current(0))?;
+            sec_offsets.insert(sec_type, offset);
+            sec_sizes.insert(sec_type, sec_size);
+            reader.seek(SeekFrom::Current(sec_size as i64))?;
+        }
+
+        let header_type = 1;
+        let constraint_type = 2;
+        let wire2label_type = 3;
+
+        reader.seek(SeekFrom::Start(*sec_offsets.get(&header_type).unwrap()))?;
```

Comment:
Option Unwraps should be replaced with 'ok_or' and having a new error type added for each of them Unwraps should be replaced with 'ok_or' and having a new error type added for each of them

Reply:
How about this? https://github.com/gakonst/ark-circom/pull/13/commits/9f22ea3625bdd9220a3412a478ff12f5c63375fc